File: fe/mobile/android/app/src/main/java/com/capstone/bridge/AuthBridgeModule.java
package com.capstone.bridge;

import com.capstone.domain.model.UserProfile;
import com.capstone.domain.repository.AuthRepository;
import com.facebook.react.bridge.*;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AuthBridgeModule extends ReactContextBaseJavaModule {
    private final AuthRepository authRepository;
    private final ExecutorService executorService;

    public AuthBridgeModule(ReactApplicationContext reactContext, AuthRepository authRepository) {
        super(reactContext);
        this.authRepository = authRepository;
        this.executorService = Executors.newFixedThreadPool(4);
    }

    @Override
    public String getName() {
        return "AuthModule";
    }

    @ReactMethod
    public void login(final String accessToken, final Promise promise) {
        executorService.execute(() -> {
            try {
                UserProfile profile = authRepository.login(accessToken);
                WritableMap map = Arguments.createMap();
                map.putString("fullName", profile.getFullName());
                map.putString("avatarUrl", profile.getAvatarUrl());
                map.putString("role", profile.getRole());
                map.putString("username", profile.getUsername());
                map.putString("email", profile.getEmail());
                promise.resolve(map);
            } catch (Exception e) {
                promise.reject("LOGIN_ERROR", e.getMessage(), e);
            }
        });
    }

    @ReactMethod
    public void sendOtp(final String email, final Promise promise) {
        executorService.execute(() -> {
            try {
                String message = authRepository.sendOtp(email);
                promise.resolve(message);
            } catch (Exception e) {
                promise.reject("SEND_OTP_ERROR", e.getMessage(), e);
            }
        });
    }

    @ReactMethod
    public void verifyOtp(final String email, final String otp, final Promise promise) {
        executorService.execute(() -> {
            try {
                String message = authRepository.verifyOtp(email, otp);
                promise.resolve(message);
            } catch (Exception e) {
                promise.reject("VERIFY_OTP_ERROR", e.getMessage(), e);
            }
        });
    }

    @ReactMethod
    public void resetPassword(final String email, final String otp, final String newPassword, final Promise promise) {
        executorService.execute(() -> {
            try {
                String message = authRepository.resetPassword(email, otp, newPassword);
                promise.resolve(message);
            } catch (Exception e) {
                promise.reject("RESET_PASSWORD_ERROR", e.getMessage(), e);
            }
        });
    }

    @ReactMethod
    public void getMe(final Promise promise) {
        executorService.execute(() -> {
            try {
                UserProfile profile = authRepository.getMe();
                WritableMap map = Arguments.createMap();
                map.putString("fullName", profile.getFullName());
                map.putString("avatarUrl", profile.getAvatarUrl());
                map.putString("role", profile.getRole());
                map.putString("username", profile.getUsername());
                map.putString("email", profile.getEmail());
                promise.resolve(map);
            } catch (Exception e) {
                promise.reject("GET_ME_ERROR", e.message, e);
            }
        });
    }
}

File: fe/mobile/android/app/src/main/java/com/capstone/bridge/AuthBridgeModule.kt
package com.capstone.bridge

import com.capstone.domain.repository.AuthRepository
import com.facebook.react.bridge.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class AuthBridgeModule(
    reactContext: ReactApplicationContext,
    private val authRepository: AuthRepository
) : ReactContextBaseJavaModule(reactContext) {

    private val scope = CoroutineScope(Dispatchers.IO)

    override fun getName(): String = "AuthModule"

    /**
     * Thực hiện đăng nhập bằng Access Token từ Keycloak.
     * Luồng: JavaScript -> AuthModule.login -> Repository.login -> Backend -> Trả về Profile
     */
    @ReactMethod
    fun login(accessToken: String, promise: Promise) {
        scope.launch {
            try {
                val profile = authRepository.login(accessToken)
                val map = Arguments.createMap().apply {
                    putString("fullName", profile.fullName)
                    putString("avatarUrl", profile.avatarUrl)
                    putString("role", profile.role)
                    putString("username", profile.username)
                    putString("email", profile.email)
                }
                promise.resolve(map)
            } catch (e: Exception) {
                promise.reject("LOGIN_ERROR", e.message, e)
            }
        }
    }

    /**
     * Gửi yêu cầu mã OTP về email để thực hiện quên mật khẩu.
     * Luồng: JavaScript -> AuthModule.sendOtp -> Repository.sendOtp -> Backend -> Gửi Email
     */
    @ReactMethod
    fun sendOtp(email: String, promise: Promise) {
        scope.launch {
            try {
                val message = authRepository.sendOtp(email)
                promise.resolve(message)
            } catch (e: Exception) {
                promise.reject("SEND_OTP_ERROR", e.message, e)
            }
        }
    }

    /**
     * Xác thực mã OTP mà người dùng nhập vào.
     * Luồng: JavaScript -> AuthModule.verifyOtp -> Repository.verifyOtp -> Backend -> Kiểm tra OTP
     */
    @ReactMethod
    fun verifyOtp(email: String, otp: String, promise: Promise) {
        scope.launch {
            try {
                val message = authRepository.verifyOtp(email, otp)
                promise.resolve(message)
            } catch (e: Exception) {
                promise.reject("VERIFY_OTP_ERROR", e.message, e)
            }
        }
    }

    /**
     * Đặt lại mật khẩu mới sử dụng kèm mã OTP đã xác thực.
     * Luồng: JavaScript -> AuthModule.resetPassword -> Repository.resetPassword -> Backend -> Cập nhật Pass
     */
    @ReactMethod
    fun resetPassword(email: String, otp: String, newPassword: String, promise: Promise) {
        scope.launch {
            try {
                val message = authRepository.resetPassword(email, otp, newPassword)
                promise.resolve(message)
            } catch (e: Exception) {
                promise.reject("RESET_PASSWORD_ERROR", e.message, e)
            }
        }
    }

    /**
     * Lấy thông tin cá nhân của người dùng hiện tại dựa trên token.
     * Luồng: JavaScript -> AuthModule.getMe -> Repository.getMe -> Backend -> Trả về Profile
     */
    @ReactMethod
    fun getMe(accessToken: String, promise: Promise) {
        scope.launch {
            try {
                val profile = authRepository.getMe(accessToken)
                val map = Arguments.createMap().apply {
                    putString("fullName", profile.fullName)
                    putString("avatarUrl", profile.avatarUrl)
                    putString("role", profile.role)
                    putString("username", profile.username)
                    putString("email", profile.email)
                }
                promise.resolve(map)
            } catch (e: Exception) {
                promise.reject("GET_ME_ERROR", e.message, e)
            }
        }
    }
}
